### PR TITLE
feat: Fetching `chain_id`

### PR DESCRIPTION
**Pull request by TAdev0**
Resolves: #673 

# Pull Request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [ ] Testing

# What is the new behavior?

chain_id is fetched using kakarot client instead of hardcoded 

# Does this introduce a breaking change?

- [ ] Yes
- [x] No

- - -
This change is [Reviewable](https://reviewable.io/reviews/kkrt-labs/kakarot-rpc/687)

### CODE REVIEW

**Comment by TAdev0** [comment]
@Eikix  a few comments : 

in src/starknet_client/mod.rs : 
```
        let chain_id = request.chain_id.unwrap();
```  will just panic if the request fails. Should i still use `unwrap_or_else` with a hardcoded value?


in tests/test_utils/evm_contract.rs , i don't see how to grab chain_id , as there is not kakarot client, and therefore no starknet_provider to use. Currently, i removed the hardcoded chain_id and set chain_id to 0 in this file. What would be the best option?

File: src/eth_rpc/servers/eth_rpc.rs [hunk 1]
```diff
@@ -77,7 +77,11 @@ impl<P: Provider + Send + Sync + 'static> EthApiServer for KakarotEthRpc<P> {
     }
 
     async fn chain_id(&self) -> Result<Option<U64>> {
-        Ok(Some(CHAIN_ID.into()))
+        let chain_id_as_field = self.kakarot_client.starknet_provider().chain_id().await.map_err(EthApiError::from)?;
```
Comment (Eikix) [inline src/eth_rpc/servers/eth_rpc.rs:80]: in Rust, it's idiomatic to shadow variable declarations, unlike Typescript, where you can't.

That is:
```rust
let chain_id = self.kakarot_client.starknet_provider().chain_id().await.map_err(EthApiError::from)?;

let chain_id = u64::try_from(chain_id).ok().unwrap;
```

Comment (Eikix) [reply src/eth_rpc/servers/eth_rpc.rs:80]: Another approach, depending on taste is to use scope:
https://doc.rust-lang.org/rust-by-example/variable_bindings/scope.html

```rust
let chain_id = {
let chain_id = self.kakarot_client.starknet_provider().chain_id().await.map_err(EthApiError::from)?;
let chain_id = u64::try_from(chain_id).ok().unwrap;
U64([chain_id_as_u64])
}
```

File: src/eth_rpc/servers/eth_rpc.rs [hunk 2]
```diff
@@ -77,7 +77,11 @@ impl<P: Provider + Send + Sync + 'static> EthApiServer for KakarotEthRpc<P> {
...
+        let chain_id_as_u64 = u64::try_from(chain_id_as_field).ok().unwrap();
```
Comment (Eikix) [inline src/eth_rpc/servers/eth_rpc.rs:81]: You shouldn't unwrap here but cast this error to a EthApiError::ConversionError

File: src/eth_rpc/servers/net_rpc.rs [hunk 3]
```diff
@@ -23,8 +22,12 @@ impl<P: Provider + Send + Sync> NetRpc<P> {
 
 #[async_trait]
 impl<P: Provider + Send + Sync + 'static> NetApiServer for NetRpc<P> {
-    fn version(&self) -> Result<U64> {
-        Ok(CHAIN_ID.into())
+    async fn version(&self) -> Result<U64> {
```
Comment (Eikix) [inline src/eth_rpc/servers/net_rpc.rs:25]: can you add a small rustdoc comment here saying net_version method returns the network ID

File: tests/test_utils/evm_contract.rs [hunk 4]
```diff
@@ -79,7 +78,7 @@ impl EvmContract for KakarotEvmContract {
         };
 
         Ok(Transaction::Eip1559(TxEip1559 {
-            chain_id: CHAIN_ID,
+            chain_id: 0,
```
Comment (Eikix) [inline tests/test_utils/evm_contract.rs:81]: why is chain_id 0?

File: src/starknet_client/mod.rs [hunk 5]
```diff
@@ -367,7 +367,7 @@ impl<P: Provider + Send + Sync> KakarotClient<P> {
             }
         };
 
-        let chain_id = request.chain_id.unwrap_or_else(|| CHAIN_ID.into());
+        let chain_id = request.chain_id.unwrap();
```
Comment (Eikix) [inline src/starknet_client/mod.rs:370]: No unwraps, we should handle the error, let's error out if the chain_id isn't provided

File: src/eth_rpc/servers/eth_rpc.rs [hunk 6]
```diff
@@ -77,7 +77,11 @@ impl<P: Provider + Send + Sync + 'static> EthApiServer for KakarotEthRpc<P> {
...
+        let chain_id_as_u64_struct = U64([chain_id_as_u64]);
```
[the remark this answers is not in the document]
Comment (TAdev0) [reply src/eth_rpc/servers/eth_rpc.rs:82]: true, thank you!

File: src/eth_rpc/servers/net_rpc.rs (hunk 3 above)
Comment (TAdev0) [reply src/eth_rpc/servers/net_rpc.rs:25]: done

File: src/eth_rpc/servers/eth_rpc.rs (hunk 1 above)
Comment (TAdev0) [reply src/eth_rpc/servers/eth_rpc.rs:80]: thank you!

**Comment by TAdev0** [comment]
@Eikix  sorry this completely forgot about this PR.
I made some changes, tell me if this is correct. I'm really new with Rust.
I still need to refactor

**Comment by TAdev0** [comment]
@Eikix  could i have a hint about how to do this refactoring? i'm not sure how to proceed

**Pull request closed without merging**